File: sum_root_to_leaf_numbers.py
class treeNode:
    def __init__(self,x):
        self.val=x
        self.left=None
        self.right=None
        self.next=None
class Solution:
    def sumNumbers(self,root):
        self.summ=0
        ans=0
        self.dfs(root,ans,self.summ)
        return self.summ
    def dfs(self,root,ans,summ):
        if not root:
            self.summ+=ans
            ans=0
            return 
        if root:
            if not root.left and not root.right:
                self.summ+=ans+root.val
                ans=0
                return
        if root.left:
            self.dfs(root.left,10*(ans+root.val),summ)
        if root.right:
            self.dfs(root.right,10*(ans+root.val),summ)
        
        
    def preorder(self,root):
        if root:
            print root.val
            self.preorder(root.left)
            self.preorder(root.right)
   
if __name__=='__main__':
    root=tree=treeNode(9)
   #tree.left=treeNode(9)
    #tree.left.right=treeNode(1)
    #tree.left.left=treeNode(11)
    #tree.left.left.left=treeNode(7)
    #tree.left.left.right=treeNode(2)
    #tree.right=treeNode(0)
    #tree.right.left=treeNode(13)
    #tree.right.right=treeNode(4)
    #tree.right.right.right=treeNode(1)    
    sol=Solution()
    print sol.sumNumbers(root)

File: Algorithm/CPP/sum_root_to_leaf_numbers.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
# include <string>
# include <unordered_map>
# include <cctype>
# include <algorithm>
# include <time.h>
# include <queue>
using namespace std;

struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
	int sumNumbers(TreeNode* root)
	{
		int sum = 0;
		if (root)		dfs(root, sum, 0);
		return sum;
	}
	void dfs(TreeNode* root, int& sum, int ans)
	{
		ans = ans * 10 + root->val;
		if (!root->left && !root->right)
		{
			sum += ans;
			return;
		}
		if (root->left)			dfs(root->left, sum, ans);
		if (root->right)		dfs(root->right, sum, ans);
	}

	void printTree(TreeNode* root)
	{
		if (root)
		{
			printTree(root->left);
			cout << root->val;
			printTree(root->right);
		}
	}
};
int  main()
{
	Solution sol;	
	TreeNode *root1;
	root1 = new TreeNode(4);
	root1->left =new  TreeNode(2);
	root1->right = new TreeNode(5);
	root1->left->left = new TreeNode(1);
	root1->left->right = new TreeNode(3);

	cout << sol.sumNumbers(root1)<< endl;
	//sol.printTree(root1);
	cout<< endl;
	system("pause");
	return 0;
}


